drivers: dma : stm32 with dmamux has a special offset

FRASTM opened the pull request:
This new offset value in the dma config device tree is made to build the table of dma mux_channels with a dmamux.
In case of DMAMUX, for each dma instance, the mux channel must reflect, thanks to this _offset_, the position of each dma channel in the table of dmamux channels.

For the stm32wb55 soc which has 2 dma instances of 7 channels each plus a dmamux of 14 channels, the DMAMUX is used with DMA1 and DMA2:
. DMAMUX channels **0** to 6 are connected to DMA1 channels 1 to 7
. DMAMUX channels **7** to 13 are connected to DMA2 channels 1 to 7
In this example the offset for the DMA1 table is **0**, offset for the DMA2 table is **7**

This PR completes the PR #28182

Signed-off-by: Francois Ramu <francois.ramu@st.com>

lowlander commented on dts/bindings/dma/dma-controller.yaml:
```diff
@@ -33,3 +33,13 @@ properties:
+        for 1st dma instance offset is 0
+        for 2nd dma instance, offset is the nb of dma channels of the 1st dma instance
+        for 3rd dma instance, offset is the nb of dma channels of the 1st
+        dma instance plus the nb of dma channels of the 2nd dma instance, etc
```
Is `dma-offset` really the right name?  Would `dmamux-offset` maybe be better? Or `dmamux-channel-offset` ?

FRASTM replied:
it could be more confusing to include dmamux in the dma binding

lowlander replied:
Well "confusing" is the right word to describe the stm32dma :-) 

I just thought it would match the code `config->streams[i].mux_channel = i + config->offset;` more closely because there it is also the mux_channel.

lowlander replied:
Or another thought, would a `dma-offset = <0, 7>;` in the dmamux driver be a solution? Although even if it is, it probably just reverses the problem you trying to fix, because than you'll need the dmamux device in the dma driver.

FRASTM replied:
I tried this or something similar : to include in the dmamux DTS, the references of the dma offset (though a phandle of dma-masters = <&dma, &dma2> or  directly the offset of each dma). However, the table is set in the **dma init**, at a time the dmamux is not yet known init'ed. 
There is a cross dependency from each dma  channels (dma_stm32_stream) [from 0 to 6]  being link to a dmamux channel [from 0 to 13] and the dmamux channels (dmamux_stm32_channel) referring to their dma instance. There is no simple config to pass info through. 
With this dma-offset in the dma DT, the value is defined where it is used.

lowlander replied:
Yeah that is what I thought, so this is kind of the the best solution that can be done. The name is no show stopper for me,  even though I think that a name like `dmamux-channel-offset` makes it clearer that it has something to do with the dmamux. But I leave that up to you, closing this review as resolved.

ABOSTM replied:
I fell like @lowlander that `dmamux-channel-offset` would be better to indicate this offset is linked to dmamux, and implicitly developer can found information in DMAMUX chapter (and not in DMA chapter).

erwango commented on dts/bindings/dma/dma-controller.yaml:
```diff
@@ -33,3 +33,13 @@ properties:
       required: false
       description: Number of DMA request signals supported by the controller.
+
+    dma-offset:
```
Since it is stm32 specific, it has to be moved to `dts/bindings/dma/st,stm32-dma.yaml`

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -650,6 +645,7 @@ const struct dma_stm32_config dma_stm32_config_##index = {		\
 	.support_m2m = DT_INST_PROP(index, st_mem2mem),			\
 	.max_streams = DMA_STM32_##index##_STREAM_COUNT,		\
 	.streams = dma_stm32_streams_##index,				\
+	.offset = DT_INST_PROP_OR(index, dma_offset, 0),	\
```
Nit: fix end of line (`\`) indentation

FRASTM replied:
done

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_wb55rg.overlay:
```diff
@@ -12,10 +12,12 @@
 
 &dma1 {
 	status = "okay";
+	dma-offset = <0>;
```
This is soc dependent, should be moved to `dts/arm/st`

FRASTM replied:
moved to soc, to be applied on each stm32 that enables its dmamux

FRASTM commented:
in the `dma_stm32_config` structure, the `offset` now depends on the CONFIG_DMAMUX_STM32
through a conditional macro of the `DMA_STM32_INIT_DEV`

lowlander commented:
@FRASTM , @erwango what is the status of this PR ?

erwango commented:
> @FRASTM , @erwango what is the status of this PR ?

@lowlander, we'll try to get it in by next week.

FRASTM commented:
rebase

FRASTM committed:
drivers: dma : stm32 with dmamux has a special offset

This new offset value in the dma config is made to
build the table of dma mux_channels with a dmamux.
Range depends on the nb of channels for selected dma instance

Signed-off-by: Francois Ramu <francois.ramu@st.com>

lowlander commented:
> > @FRASTM , @erwango what is the status of this PR ?
> 
> @lowlander, we'll try to get it in by next week.

@erwango  is it already next week ? ;-)

erwango commented:
> @erwango is it already next week ? ;-)

Sorry @lowlander, there has been some delay indeed. @FRASTM can you have a check ?

FRASTM committed:
dts: dma : stm32 soc with dmamux has a special offset

The offset value in the dma config is required when the
soc has dmamux like the stm32wb series

Signed-off-by: Francois Ramu <francois.ramu@st.com>
